admin/win/tools/NCMsiHelper/NCMsiHelper.cpp

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) by Michael Schuster <michael.schuster@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "NCTools.h"
+#include "utility.h"
+#include "LogResult.h"
+#include "NCMsiHelper.h"
+
+using namespace NCTools;
+
+HRESULT NCMSIHELPER_API DoExecNsisUninstaller(int argc, LPWSTR *argv)
+{
+    if (argc != 2) {
+        return HRESULT_FROM_WIN32(ERROR_INVALID_PARAMETER);
+    }
+
+    auto appShortName = std::wstring(argv[0]);
+    auto uninstallExePath = std::wstring(argv[1]);
+
+    if (appShortName.empty()
+         || uninstallExePath.empty()) {
+        return HRESULT_FROM_WIN32(ERROR_INVALID_PARAMETER);
+    }
+
+    auto appInstallDir = uninstallExePath;
+    auto posLastSlash = appInstallDir.find_last_of(PathSeparator);
+    if (posLastSlash != std::wstring::npos) {
+        appInstallDir.erase(posLastSlash);
+    } else {
+        return HRESULT_FROM_WIN32(ERROR_INVALID_PARAMETER);
+    }
+
+    // Run uninstaller
+    std::wstring cmd = L'\"' + uninstallExePath + L"\" /S _?=" + appInstallDir;
+    LogResult(S_OK, "Running '%ls'.", cmd.data());
+    Utility::execCmd(cmd);
+
+    LogResult(S_OK, "Waiting for NSIS uninstaller.");
+
+    // Can't wait for the process because Uninstall.exe (opposed to Setup.exe) immediately returns, so we'll sleep a bit.
+    Utility::waitForNsisUninstaller(appShortName);
+
+    LogResult(S_OK, "Removing the NSIS uninstaller.");
+
+    // Sleep a bit and clean up the NSIS mess
+    Sleep(1500);
```

Comment:
That's already balanced to be enough time ;p

The uninstaller needs a few 100 msecs to close its EXE handle, so the file can then be deleted. Otherwise deletion fails and the uninstall.exe sticks around in the new MSI program folder for eternity, except there is a statement in the WiX file to delete it, what feels quite ugly somehow.